blacktie/jatmibroker-xatmi/src/main/java/org/jboss/narayana/blacktie/jatmibroker/core/transport/hybrid/stomp/StompManagement.java

```diff
@@ -67,28 +67,30 @@ public static Socket connect(String host, int port, String username, String pass
     }
 
     public static void send(Message message, OutputStream outputStream) throws IOException {
-        log.trace("Writing on: " + outputStream);
-        StringBuffer toSend = new StringBuffer(message.getCommand().toString());
-        toSend.append("\n");
-
-        Iterator<String> keys = message.getHeaders().keySet().iterator();
-        while (keys.hasNext()) {
-            String key = keys.next();
-            String value = message.getHeaders().get(key);
-            toSend.append(key);
-            toSend.append(":");
-            toSend.append(value);
+        synchronized (outputStream) {
```

Comment:
For example the javadoc for java.nio.channels.SocketChannel says:

"Socket channels are safe for use by multiple concurrent threads. They support concurrent reading and writing, though at most one thread may be reading and at most one thread may be writing at any given time. The connect and finishConnect methods are mutually synchronized against each other, and an attempt to initiate a read or write operation while an invocation of one of these methods is in progress will block until that invocation is complete."

The javadoc for java.net.Socket isn't as definitive but I would expect/hope for something similar or you could use NIO instead.